codex/stores/repostore.nim

```diff
@@ -41,119 +45,277 @@ declareGauge(codex_repostore_bytes_reserved, "codex repostore bytes reserved")
-  self: RepoStore,
-  treeCid: Cid,
-  index: Natural,
-  blockCid: Cid,
-  proof: MerkleProof
-): Future[?!void] {.async.} =
-  ## Put a block to the blockstore
-  ##
+proc encode(t: QuotaUsage): seq[byte] = t.autoencode
+proc decode(T: type QuotaUsage, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+proc encode(t: BlockMetadata): seq[byte] = t.autoencode
+proc decode(T: type BlockMetadata, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+proc encode(t: LeafMetadata): seq[byte] = t.autoencode
+proc decode(T: type LeafMetadata, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+###########################################################
+# Helper types and procs
+###########################################################
+
+type
+  DeleteResultKind = enum
+    Deleted = 0,    # block removed from store
+    InUse = 1,      # block not removed, refCount > 0 and not expired
+    NotFound = 2    # block not found in store
+
+  DeleteResult = object
+    kind: DeleteResultKind
+    released: NBytes
+
+  StoreResultKind = enum
+    Stored = 0,         # new block stored
+    AlreadyInStore = 1  # block already in store
+
+  StoreResult = object
+    kind: StoreResultKind
+    used: NBytes
+
+proc encode(t: DeleteResult): seq[byte] = t.autoencode
+proc decode(T: type DeleteResult, bytes: seq[byte]): ?!T = T.autodecode(bytes)
+
+proc encode(t: StoreResult): seq[byte] = t.autoencode
+proc decode(T: type StoreResult, bytes: seq[byte]): ?!T = T.autodecode(bytes)
 
+proc putLeafMetadata(self: RepoStore, treeCid: Cid, index: Natural, blkCid: Cid, proof: MerkleProof): Future[?!StoreResultKind] {.async.} =
   without key =? createBlockCidAndProofMetadataKey(treeCid, index), err:
     return failure(err)
 
-  let value = (blockCid, proof).encode()
+  await self.metaDs.modifyTGetU(key,
+    proc (maybeCurrMd: ?LeafMetadata): Future[(?LeafMetadata, StoreResultKind)] {.async.} =
+      var
+        md: LeafMetadata
+        res: StoreResultKind
 
-  await self.metaDs.put(key, value)
+      if currMd =? maybeCurrMd:
+        md = currMd
+        res = AlreadyInStore
+      else:
+        md = LeafMetadata(blkCid: blkCid, proof: proof)
+        res = Stored
 
-proc getCidAndProof(
-  self: RepoStore,
-  treeCid: Cid,
-  index: Natural
-): Future[?!(Cid, MerkleProof)] {.async.} =
+      (md.some, res)
+  )
+
+proc getLeafMetadata(self: RepoStore, treeCid: Cid, index: Natural): Future[?!LeafMetadata] {.async.} =
   without key =? createBlockCidAndProofMetadataKey(treeCid, index), err:
     return failure(err)
 
-  without value =? await self.metaDs.get(key), err:
+  without leafMd =? await self.metaDs.get(key), err:
     if err of DatastoreKeyNotFound:
       return failure(newException(BlockNotFoundError, err.msg))
     else:
       return failure(err)
 
-  return (Cid, MerkleProof).decode(value)
+  LeafMetadata.decode(leafMd)
+
+proc updateTotalBlocksCount(self: RepoStore, plusCount: Natural = 0, minusCount: Natural = 0): Future[?!void] {.async.} =
```

Comment:
>Give given how much of a pain all the type conversions are becoming, I'm not sure we should change this.

That's exactly the reason. My initial implementation was to use single `deltaCount: int` parameter. However in the end it's actually waay easier to avoid conversion errors and arithmetic overflow/underflow errors if we have two parameters. We could also go with `amount: Nautral, op: Add | Subtract` but that's still 2 params.